README.md
<p align="center">
  <img 
    src="http://res.cloudinary.com/vidsy/image/upload/v1503160820/CoZ_Icon_DARKBLUE_200x178px_oq0gxm.png" 
    width="125px"
  >
</p>

<h1 align="center">neo-go-sdk</h1>

<p align="center">
  Golang SDK for the <b>NEO</b> blockchain.
</p>

<p align="center">
  <a href="https://github.com/CityOfZion/neo-go-sdk/releases">
    <img src="https://img.shields.io/github/tag/CityOfZion/neo-go-sdk.svg?style=flat">
  </a>
  <a href="https://circleci.com/gh/CityOfZion/neo-go-sdk/tree/master">
    <img src="https://circleci.com/gh/CityOfZion/neo-go-sdk/tree/master.svg?style=shield">
  </a>
  <a href="https://goreportcard.com/report/github.com/CityOfZion/neo-go-sdk">
    <img src="https://goreportcard.com/badge/github.com/CityOfZion/neo-go-sdk">
  </a>
  <a href="https://godoc.org/github.com/CityOfZion/neo-go-sdk/neo">
    <img src="https://godoc.org/github.com/CityOfZion/neo-go-sdk?status.svg">
  </a>
</p>

## What?

- Client for interacting with a node on the [NEO](http://neo.org/) blockchain.
- Retrieve data and send actions.
- Fully tested [Golang](https://golang.org/) package.
- Aimed to help other developers build applications for the NEO ecosystem.
- Written using the standard library, without 3rd party packages. 

## Quick Start

```
go get github.com/CityOfZion/neo-go-sdk
```

```golang
package main

import (
  "log"

  "github.com/CityOfZion/neo-go-sdk/neo"
)

func main() {
  nodeURI := "http://test1.cityofzion.io:8880"
  client := neo.NewClient(nodeURI)

  ok := client.Ping()
  if !ok {
    log.Fatal("Unable to connect to NEO node")
  }

  block, err := client.GetBlockByHash(
    "3f0b498c0d57f73c674a1e28045f5e9a0991f9dac214076fadb5e6bafd546170",
  )
  if err != nil {
    log.Fatal(err)
  }

  log.Printf("Block found, index is: %d", block.Index)
}
```

## Examples

See [GoDoc](https://godoc.org/github.com/CityOfZion/neo-go-sdk/neo) for full documentation.

## CLI

![https://res.cloudinary.com/vidsy/image/upload/v1506873093/Oct-01-2017_16-50-54_j4j3hp.gif](https://res.cloudinary.com/vidsy/image/upload/v1506873093/Oct-01-2017_16-50-54_j4j3hp.gif)

Debugging a NEO public and private key pair is a common task when interacting with the blockchain.
Make use of the **neo-go-sdk** CLI to help with this process:

```
./neo-go-sdk --wif KxQREAjBL6Ga8dw9rPN45pwoZ5dxhAQacEajQke6qmpB7DW6nAWE
```

This will output the **full details** about the key pair. See [releases](https://github.com/CityOfZion/neo-go-sdk/releases) to download the CLI.

## Help

- Open a new [issue](https://github.com/CityOfZion/neo-go-sdk/issues/new) if you encountered a problem.
- Or ping **@revett** on the [NEO Slack](https://neo-slack-invite.herokuapp.com).
- Submitting PRs to the project is always welcome! 🎉
- Check the [Changelog](https://github.com/CityOfZion/neo-go-sdk/blob/master/CHANGELOG.md) for recent changes.

## License

- Open-source [MIT](https://github.com/CityOfZion/neo-go-sdk/blob/master/LICENSE).
- Main author is [@revett](https://github.com/revett).
- This project adheres to the [Contributor Covenant Code of Conduct](https://github.com/goreleaser/goreleaser/blob/master/CODE_OF_CONDUCT.md).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this
project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## v1.9.1 - 2018-01-22

### Fixed

- [#23](https://github.com/CityOfZion/neo-go-sdk/issues/23) fixed, now returns an error from `ModularArithmetic#Sqrt`.

## v1.9.0 - 2018-01-22

### Added

- New function to select the best node from a array of node URIs.

```golang
client, err := neo.NewClientUsingMultipleNodes(
  []string{
    "http://seed1.neo.org:10332",
    "http://seed2.neo.org:10332",
    "http://seed3.neo.org:10332",
  },
)

err = client.SelectBestNode()
```

## v1.8.0 - 2017-10-27

### Added

- New function to validate a public NEO address.

```golang
isValid, err := client.ValidateAddress("ARnLq3m1jsrZU7SS7jLHAvNm37GmaZbsPy")
```

## v1.7.0 - 2017-10-27

### Added

- New function to fetch a storage value of a given smart contract.

```golang
storage, err := client.GetStorage(
  "0xecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9", // RPX smart contract
  "totalSupply",
)
```

## v1.6.1 - 2017-10-13

### Changed

- Updated client unit tests to adhere to the new JSON RPC hash representations.
- Performed some general cleanup and organization in existing client unit tests.

## v1.6.0 - 2017-09-01

### Added

- Ability to derive signature from private key.
- CLI outputs signature from WIF using new function.

```golang
privateKey, err := neo.NewPrivateKeyFromWIF("L1QqQJnpBwbsPGAuutuzPTac8piqvbR1HRjrY5qHup48TBCBFe4g")
if err != nil {
  log.Fatal(err)
}

signatureBytes, err := privateKey.Signature()
if err != nil {
  log.Fatal(err)
}

signature := hex.EncodeToString(signatureBytes)

log.Println(signature)
```

## v1.5.0 - 2017-09-01

### Added

- Ability to derive public key value from private key.
- CLI outputs public key from WIF using new function.

```golang
privateKey, err := neo.NewPrivateKeyFromWIF("L1QqQJnpBwbsPGAuutuzPTac8piqvbR1HRjrY5qHup48TBCBFe4g")
if err != nil {
  log.Fatal(err)
}

publicKeyBytes, err := privateKey.PublicKey()
if err != nil {
  log.Fatal(err)
}

publicKey := hex.EncodeToString(publicKeyBytes)

log.Println(publicKey)
```

## v1.4.1 - 2017-09-01

### Added

- GIF to README to help with documenting the new CLI.

## Changed

- Add `dist/` to **.gitignore**.

## v1.4.0 - 2017-08-19

### Added

- Documentation and configuration for the CLI.

## v1.3.0 - 2017-08-19

### Added

- CLI for debugging a NEO public and private key pair (see [releases](https://github.com/CityOfZion/neo-go-sdk/releases)):

```
./neo-go-sdk --wif KxQREAjBL6Ga8dw9rPN45pwoZ5dxhAQacEajQke6qmpB7DW6nAWE
```

## v1.2.0 - 2017-08-19

### Added

- [Elliptic curve](https://en.wikipedia.org/wiki/Elliptic_curve) implementation in utility package.
- Base58 encoding support.
- Migrated `neo.WIF` struct into `neo.PrivateKey` struct.
- Derive a public NEO address from a private key (WIF):

```golang
privateKey, err := neo.NewPrivateKeyFromWIF("L1QqQJnpBwbsPGAuutuzPTac8piqvbR1HRjrY5qHup48TBCBFe4g")
if err != nil {
  log.Fatal(err)
}

publicAddress, err := privateKey.PublicAddress()
if err != nil {
  log.Fatal(err)
}

log.Println(publicAddress)
```

## v1.1.2 - 2017-08-12

### Changed

- TCP connections made by `client.Ping()` are now closed to stop memory leaks from happening.

## v1.1.1 - 2017-08-12

### Changed

- **@eramus** fixed:
  - closing response body in wrong place.
  - made Base58 decode much more efficient and clean.
- Slack URI in README was updated.

## v1.1.0 - 2017-08-19

### Added

- New `neo.WIF` struct.
- New `neo.PrivateKey` struct.
- Ability to convert a WIF into a PrivateKey, see example below.

```golang
wif := neo.NewWIF("L1QqQJnpBwbsPGAuutuzPTac8piqvbR1HRjrY5qHup48TBCBFe4g")

privateKey, err := wif.ToPrivateKey()
if err != nil {
  log.Fatal(err)
}

log.Println(privateKey.Value)
```

## v1.0.1 - 2017-08-19

### Added

- Link to GoDoc documentation.

### Changed

- Logo in README to new CoZ logo.

## v1.0.0 - 2017-08-16

### Changed

- Added badges to README.

## v0.2.0 - 2017-08-16

### Added

- Existing code from original repo.
- Full CI job.

## v0.1.0 - 2017-08-16

### Added

- Setup repo.


utility/base_58.go
package utility

import (
	"fmt"
	"math/big"
)

type (
	// Base58 is a encode/decode utility for base58 strings.
	Base58 struct {
		decodeMap map[rune]int64
	}
)

const prefix rune = '1'

var decodeMap = map[rune]int64{
	'1': 0, '2': 1, '3': 2, '4': 3, '5': 4,
	'6': 5, '7': 6, '8': 7, '9': 8, 'A': 9,
	'B': 10, 'C': 11, 'D': 12, 'E': 13, 'F': 14,
	'G': 15, 'H': 16, 'J': 17, 'K': 18, 'L': 19,
	'M': 20, 'N': 21, 'P': 22, 'Q': 23, 'R': 24,
	'S': 25, 'T': 26, 'U': 27, 'V': 28, 'W': 29,
	'X': 30, 'Y': 31, 'Z': 32, 'a': 33, 'b': 34,
	'c': 35, 'd': 36, 'e': 37, 'f': 38, 'g': 39,
	'h': 40, 'i': 41, 'j': 42, 'k': 43, 'm': 44,
	'n': 45, 'o': 46, 'p': 47, 'q': 48, 'r': 49,
	's': 50, 't': 51, 'u': 52, 'v': 53, 'w': 54,
	'x': 55, 'y': 56, 'z': 57,
}

// NewBase58 creates a new Base58 struct, using the default alphabet.
func NewBase58() Base58 {
	base58 := Base58{}
	return base58
}

// Decode decodes the base58 encoded string.
func (b Base58) Decode(s string) ([]byte, error) {
	startIndex := 0
	zero := 0

	for i, c := range s {
		if c == prefix {
			zero++
		} else {
			startIndex = i
			break
		}
	}

	n := big.NewInt(0)
	div := big.NewInt(58)

	for _, c := range s[startIndex:] {
		charIndex, ok := decodeMap[c]
		if !ok {
			return nil, fmt.Errorf(
				"invalid character '%c' when decoding this base58 string: '%s'", c, s,
			)
		}

		n.Add(n.Mul(n, div), big.NewInt(charIndex))
	}

	out := n.Bytes()
	buf := make([]byte, (zero + len(out)))
	copy(buf[zero:], out[:])

	return buf, nil
}

// Encode encodes a byte slice to be a base58 encoded string.
func (b Base58) Encode(bytes []byte) string {
	lookupTable := "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

	x := new(big.Int).SetBytes(bytes)

	r := new(big.Int)
	m := big.NewInt(58)
	zero := big.NewInt(0)
	encoded := ""

	for x.Cmp(zero) > 0 {
		x.QuoRem(x, m, r)
		encoded = string(lookupTable[r.Int64()]) + encoded
	}

	return encoded
}
